core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/util/TripleSources.java

```diff
@@ -0,0 +1,277 @@
+						return (v instanceof Resource);
+					}
+				}) {
+
+			@Override
+			protected Resource convert(Value v)
+				throws QueryEvaluationException
+			{
+				return (Resource)v;
+			}
+		};
+	}
+
+	public static Iteration<? extends Value, QueryEvaluationException> list(final Resource subj,
+			final TripleSource store)
+		throws QueryEvaluationException
+	{
+		if (subj == null) {
+			throw new NullPointerException("RDF list subject cannot be null");
+		}
+		return new Iteration<Value, QueryEvaluationException>() {
+
+			Resource list = subj;
+
+			@Override
+			public boolean hasNext()
+				throws QueryEvaluationException
+			{
+				return !RDF.NIL.equals(list);
+			}
+
+			@Override
+			public Value next()
+				throws QueryEvaluationException
+			{
+				Value v = singleValue(list, RDF.FIRST, store);
+				if (v == null) {
+					throw new QueryEvaluationException("List missing rdf:first: " + list);
+				}
+				Resource nextList = (Resource)singleValue(list, RDF.REST, store);
+				if (nextList == null) {
+					throw new QueryEvaluationException("List missing rdf:rest: " + list);
+				}
+				list = nextList;
+				return v;
+			}
+
+			@Override
+			public void remove()
+				throws QueryEvaluationException
+			{
+				throw new UnsupportedOperationException();
+			}
+		};
+	}
+
+	public static boolean booleanValue(Resource subj, IRI pred, TripleSource store)
+		throws QueryEvaluationException
+	{
+		Value v = TripleSources.singleValue(subj, pred, store);
+		if (v == null) {
+			return false;
+		}
+		else if (v instanceof Literal) {
+			try {
+				return ((Literal)v).booleanValue();
+			}
+			catch (IllegalArgumentException e) {
+				throw new QueryEvaluationException(
+						"Value for " + pred + " must be of datatype " + XMLSchema.BOOLEAN + ": " + subj);
+			}
+		}
+		else {
+			throw new QueryEvaluationException("Non-literal value for " + pred + ": " + subj);
+		}
+	}
+
+	public static Value singleValue(Resource subj, IRI pred, TripleSource store)
+		throws QueryEvaluationException
+	{
+		Statement stmt = single(subj, pred, null, store);
+		return (stmt != null) ? stmt.getObject() : null;
+	}
+
+	/**
+	 * Returns the single statement with the given subject, predicate and object or null if none exists.
+	 * Context information is disregarded.
+	 * 
+	 * @param subj
+	 *        null for any.
+	 * @param pred
+	 *        null for any.
+	 * @param obj
+	 *        null for any.
+	 * @throws QueryEvaluationException
+	 *         If there is more than one such statement.
+	 */
+	public static Statement single(Resource subj, IRI pred, Value obj, TripleSource store)
+		throws QueryEvaluationException
+	{
+		Statement stmt;
+		CloseableIteration<? extends Statement, QueryEvaluationException> stmts = store.getStatements(subj,
+				pred, obj);
+		try {
+			if (stmts.hasNext()) {
+				stmt = stmts.next();
+				while (stmts.hasNext()) {
+					Statement nextStmt = stmts.next();
+					if (!org.eclipse.rdf4j.model.util.Statements.isSameTriple(stmt, nextStmt)) {
```

Comment:
This is the core of the fix as proposed by @pulquero . The main change is that instead of converting every statement to triple-only form, we simply do a different comparison.